File: test/test/ViewController.m
//
//  ViewController.m
//  test
//
//  Created by 叶常青 on 2023/11/23.
//

#import "ViewController.h"
#import <sensor/sensor.h>



@interface ViewController ()<SensorControllerDelegate>
@property (atomic, retain) SensorController* profile;
@property (atomic, strong) BLEPeripheral* device;
@property (atomic, assign) BOOL hasStartDataTransfer;
@end

const int PACKAGE_COUNT = 10;

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.profile = [[SensorController alloc] init];
    self.profile.delegate = self;
}

- (IBAction)onScan:(id)sender{
    self.deviceText.text = @"scaning";
    [self.profile startScan:5];
}

- (IBAction)onConnect:(id)sender{
    if (self.profile.state >= BLEStateConnected){
        [self.profile disconnect];
    }else if (self.device != nil){
        [self.profile connect:self.device];
    }
}

- (IBAction)onVersion:(id)sender{
    [self.profile initDataNotification:PACKAGE_COUNT cb:^(GF_RET_CODE resp) {
        if (resp == GF_SUCCESS){
            [self.profile getControllerFirmwareVersion:^(GF_RET_CODE resp, NSString *firmwareVersion) {
                dispatch_async(dispatch_get_main_queue(), ^{
                    self.versionText.text = firmwareVersion;
                });
                
            } timeout:5];
        }else{
            dispatch_async(dispatch_get_main_queue(), ^{
                self.versionText.text = @"Init fail";
            });
        }
    }];
}

- (IBAction)onDataAction:(id)sender{
    if (self.profile.state >= BLEStateConnected && self.hasStartDataTransfer){
        [self stopDataTransfer];
    }else if (self.device != nil  && !self.hasStartDataTransfer){
        [self startDataTransfer];
    }
}

-(void)startDataTransfer{
    if (![self.profile hasInitDataNotification]){
        self.dataText.text = @"please get version first";
        return;
    }
    [self.profile startDataNotification];
    self.hasStartDataTransfer = TRUE;
}

-(void)stopDataTransfer{
    [self.profile stopDataNotification];
    self.hasStartDataTransfer = FALSE;
    dispatch_async(dispatch_get_main_queue(), ^{
        self.dataText.text = @"data transfer stoped";
    });
}


#pragma mark - SynchronyDelegate
- (void)onSensorErrorCallback: (NSError*)err{
    NSLog(@"got gforce error %@", err);
}

- (void)onSensorStateChange: (BLEState)newState{
    dispatch_async(dispatch_get_main_queue(), ^{
        if (newState >= BLEStateConnected){
            self.statusText.text = @"Connected";
        }else{
            self.statusText.text = @"Not Connected";
            self.hasStartDataTransfer = FALSE;
        }
    });
}

- (void)onSensorScanResult:(NSArray *)bleDevices{
    int maxRSSI = -1000;
    BLEPeripheral* bleDeviceMax = nil;
    for (BLEPeripheral* bleDevice in bleDevices) {
        NSLog(@"Found device: %@, mac: %@, rssi: %@", bleDevice.peripheralName, bleDevice.macAddress, bleDevice.rssi);
        if ([bleDevice.rssi intValue] > maxRSSI){
            maxRSSI = [bleDevice.rssi intValue];
            bleDeviceMax = bleDevice;
        }
    }
    if (bleDeviceMax != nil){
        self.deviceText.text = [NSString stringWithFormat:@"name: %@\n mac: %@\n rssi: %@", bleDeviceMax.peripheralName, bleDeviceMax.macAddress, bleDeviceMax.rssi];
        self.device = bleDeviceMax;
    }else{
        self.deviceText.text = @"scan finish";
    }
}

- (void)onSensorNotifyData:(SensorData *)sensorData {
    NSLog(@"got data: %d", sensorData.dataType);
    NSLog(@"%@", sensorData.channelSamples);
}

@end

File: test/test2/ViewController.swift
//
//  ViewController.swift
//  test
//
//  Created by 叶常青 on 2024/4/18.
//

import UIKit
import sensor

let PACKAGE_COUNT = Int32(10)
let TIMEOUT = TimeInterval(6)

class SensorDataContext : SensorProfileDelegate{
    public var profile: SensorProfile
    public var lastEEG: SensorData?
    public var lastECG: SensorData?
    public var lastACC: SensorData?
    public var lastGYRO: SensorData?
    public var lastError: Error?
    
    init(profile: SensorProfile!) {
        self.profile = profile;
        profile.delegate = self
    }
    
    func onSensorErrorCallback(_ err: Error!) {
        lastError = err;
    }
    
    func onSensorStateChange(_ newState: BLEState) {
        print("Device: " + profile.device.name + " state: " + profile.stateString)
        if (newState == BLEState.unConnected || newState == BLEState.invalid){
            print("Reset device: " + profile.device.name);
            clear()
        }else if (newState == BLEState.ready && !profile.hasInit){
            Task{
                if (!profile.hasInit){
                    let hasInit = await profile.initAll(PACKAGE_COUNT, timeout: TIMEOUT)
                    if (hasInit){
                        print("Init " + profile.device.macAddress + " succeed");
                        print("EEG channel count:" + String(profile.eegChannelCount));
                        print("ECG channel count:" + String(profile.ecgChannelCount));
                        print("ACC channel count:" + String(profile.accChannelCount));
                        print("GYRO channel count:" + String(profile.gyroChannelCount));
                    }else{
                        print("Init " + profile.device.macAddress + " fail");
                    }
                }
            }
        }
    }
    
    func onSensorNotify(_ rawData: SensorData!) {
        if (rawData.dataType == NotifyDataType.NTF_EEG){
            print(profile.device.name + " => Got EEG data: " + String(rawData.channelSamples[0][0].timeStampInMs));

//            lastEEG?.channelSamples[0][0].timeStampInMs  : timestamp for this signal
//            lastEEG?.channelSamples[0][0].isLost         : check it and do some logic if the data is lost
//            lastEEG?.channelSamples[0][0].convertData    : physical unit is uV
//            lastEEG?.channelSamples[0][0].impedance      : this is for impedance chech
            
        }else if (rawData.dataType == NotifyDataType.NTF_ECG){
            print(profile.device.name + " => Got ECG data: " + String(rawData.channelSamples[0][0].timeStampInMs));
            lastECG = rawData
        }else if (rawData.dataType == NotifyDataType.NTF_ACC_DATA){
            print(profile.device.name + " => Got ACC data: " + String(rawData.channelSamples[0][0].timeStampInMs));
            lastACC = rawData
        }else if (rawData.dataType == NotifyDataType.NTF_GYO_DATA){
            print(profile.device.name + " => Got GYRO data: " + String(rawData.channelSamples[0][0].timeStampInMs));
            lastGYRO = rawData
        }
        
    }
    
    func clear(){
        lastEEG = nil
        lastECG = nil
        lastACC = nil
        lastGYRO = nil
        lastError = nil
    }
}

class ViewController: UIViewController , SensorControllerDelegate {
    private var controller: SensorController?
    private var sensorDataCtxs: [String : SensorDataContext] = [:]
    private var hasStartDataTransfer = false


    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        controller = SensorController.getInstance()
        controller?.delegate = self
    }

    @IBAction func onScan(_ sender: Any) {
//        deviceText.text = "scaning"
        if (!controller!.isEnable){
            print("Please open blue tooth")
            return
        }
        if (controller!.isScaning){
            controller!.stopScan()
        }else{
            controller!.startScan(TIMEOUT)
        }
        
    }

    @IBAction func onConnect(_ sender: Any) {
        for sensorData in sensorDataCtxs {
            if (sensorData.value.profile.state == BLEState.ready){
                sensorData.value.profile.disconnect()
            }else{
                sensorData.value.profile.connect()
            }
        }
    }

    @IBAction func onVersion(_ sender: Any) {
        Task{
            for sensorData in self.sensorDataCtxs {
                if (sensorData.value.profile.state == BLEState.ready){
                    let deviceInfo = await sensorData.value.profile.deviceInfo(TIMEOUT)
                    if (deviceInfo != nil){
                        print("deviceInfo: " + deviceInfo!.modelName + " : " + deviceInfo!.firmwareVersion)
                    }else{
                        print("Get deviceInfo fail: "  + sensorData.value.profile.device.name)
                    }
                
                    let battery = await sensorData.value.profile.battery(TIMEOUT)
                    if (battery >= 0){
                        print("Battery: " + sensorData.value.profile.device.name + " : " + String(battery))
                    }else{
                        print("Get battery fail: "  + sensorData.value.profile.device.name)
                    }
                }
            }
        }
    }

    @IBAction func onTest(_ sender: Any) {
        Task{
            for sensorData in self.sensorDataCtxs {
                if (sensorData.value.profile.state == BLEState.ready){
                    if (sensorData.value.profile.hasStartDataNotification){
                        sensorData.value.profile.stopDataNotification();
                    }else{
                        sensorData.value.profile.startDataNotification();
                    }
                }
            }
        }
    }
    
    func onSensorScanResult(_ bleDevices: [BLEPeripheral]) {
        print("onSensorScanResult")
        for bleDevice in bleDevices {
            if (bleDevice.name.hasPrefix("OB")){
                if (sensorDataCtxs[bleDevice.macAddress] == nil){
                    let sensorProfile = controller?.getSensor(bleDevice.macAddress);
                    let sensorDataCtx = SensorDataContext(profile : sensorProfile)
                    sensorDataCtxs[bleDevice.macAddress] = sensorDataCtx;
                    print("Found: " + bleDevice.name + " : " + String(bleDevice.rssi.intValue))
                }
            }
        }
    }
}


